src/transformers/deepspeed.py

```diff
@@ -283,36 +160,30 @@ def trainer_config_finalize(self, args, model, num_training_steps):
             )
 
 
-# keep the config object global to be able to access it anywhere during TrainingArguments life-cycle
-_hf_deepspeed_config_weak_ref = None
-
-
-def set_hf_deepspeed_config(hf_deepspeed_config_obj):
-    # this is a special weakref global object to allow us to get to Deepspeed config from APIs
-    # that don't have an easy way to get to the Deepspeed config outside of the Trainer domain.
-    global _hf_deepspeed_config_weak_ref
-    # will go away automatically when HfDeepSpeedConfig is destroyed (when TrainingArguments is destroyed)
-    _hf_deepspeed_config_weak_ref = weakref.ref(hf_deepspeed_config_obj)
-
-
-def unset_hf_deepspeed_config():
-    # useful for unit tests to ensure the global state doesn't leak - call from `tearDown` method
-    global _hf_deepspeed_config_weak_ref
-    _hf_deepspeed_config_weak_ref = None
-
-
 def is_deepspeed_zero3_enabled():
-    if _hf_deepspeed_config_weak_ref is not None and _hf_deepspeed_config_weak_ref() is not None:
-        return _hf_deepspeed_config_weak_ref().is_zero3()
-    else:
-        return False
+    if (
+        "deepspeed_plugin" in AcceleratorState._shared_state
+        and AcceleratorState._shared_state["deepspeed_plugin"] is not None
```

Comment:
Let's instantiate `AcceleratorState()` instead of looking at its `_sharde_state`.

Reply:
Hello, at first I was creating `AcceleratorState` object like you recommend in the following comments. But, then I found out that in `notebook_launcher` in `launchers.py`, `len(Accelerator._shared_state)>0` is done to throw error. As much, it will always throw error in `notebook_launcher` as calling `AutoModel.from_pretrained` will instantiate `AcceleratorState` object in the purposed changes. Also, we need to be careful with calling  `AcceleratorState()` as it was creating and intializing cuda in multiprocessing setup which would eventually throw `cuda already initialized` errors. For all these reasons, I changes from working with state object to its `_shared_state`.

Reply:
why not just copy the transformers implementation verbatim and wrap around it for accelerate needs? The transformers implementation requires no complicated objects with side-effects. Let's keep simple things simple and difficult things possible.

Also heads up, we are discussing having Deepspeed provide this functionality in its core - it may take some weeks months to happens, but it's good to plan to have this whole object management in the deepspeed side. The idea is to have the config object to be part of the deepspeed object, but not require its initialize to be able to use it. There is no API yet, but it'd be something similar to what transformer has except built into deepspeed's API.

Reply:
Hello Stas, earlier `weak_ref` was needed for checking if stage_3 is enabled or not. The effort of this PR is to get rid of `weak_ref` and `HFDeepSpeedConfig` has been copied verbatim in Accelerate. 

These changes are required to know the deepspeed stage and config without using weakref. 

It's not complicated as such as we are checking Accelerator state's content which will be same/available throughout and thereby overcoming the need for weakref.

Reply:
yes, of course, but if possible, let's keep the API simple for non-accelerator uses. When using with (1) HF Trainer or (2) w/o any trainer at all, there shouldn't be a need to create an Accelerator object since in both of these cases Accelerator is not used. So ideally there should be simple low-level API that lives in the `accelerate` codebase, but doesn't need to really rely on `accelerate` high-end functionality. I hope it makes sense. But I trust that between you and Sylvain you guys will sort it out as I don't know anything about `accelerate`.